[doc] add functional optimizer to doc

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45513 [doc] add functional optimizer to doc**

Add the basic documentation for functional optimizer to the doc

screenshot:

![Screen Shot 2020-09-30 at 2 54 51 PM](https://user-images.githubusercontent.com/9443650/94743988-1ecb8c80-032d-11eb-9ac0-fb283134638e.png)


Differential Revision: [D23995418](https://our.internmc.facebook.com/intern/diff/D23995418)

Commit:
[doc] add functional optimizer to doc

Add the basic documentation for functional optimizer to the doc

[ghstack-poisoned]

albanD commented on docs/source/optim.rst:
```diff
@@ -279,3 +279,30 @@ and start to collect SWA averages of the parameters at epoch 160:
+Functional Optimizer API
+---------------------------
+
+If you would like to update your parameters in a pure functional way, you can
```
nit this paragraph seems to repeat itself quite a bit.
I guess the idea you want here is just that this is a functional API as it will only update the given parameters.
This can be used by users that want to just perform a single step of an optimizer for a given state.

albanD reviewed (commented):
Looks good, all the ideas are there.
Maybe the phrasing could be improved but I'll let @vincentqb give his opinion on that.

vincentqb commented on docs/source/optim.rst:
```diff
@@ -279,3 +279,30 @@ and start to collect SWA averages of the parameters at epoch 160:
+optimizer allow you to only do the computation for the parameters you passed in,
+and inplace update the parameters accordingly.
+
+Note that in order to use the functional API, you should maintain the states
+for different algorithms on your own, and update some part of your non-tensor
+states before or after the call to this API (e.g. To use
+:func:`torch.optim.functional.adagrad`, `state_steps` for each parameter need
+to be updated before call to :func:`torch.optim.functional.adagrad`)
```
nit: simplifying a little
```suggestion
Note that you need to maintain the states of a functional optimizer by yourself.
In particular, you need to update some part of your non-tensor states before or
after the call to this API (for instance, `state_steps` for each parameter need to
be updated before call to :func:`torch.optim.functional.adagrad`)
```

vincentqb replied:
by the way, I do not see the instructions on how to update the states in `torch.optim.functional.adagrad` -- am i missing them?

wanchaol replied:
I added `tensor states` together with parameters in the first paragraph to indicate they are updated inplace inside the functional API. Do you think this is enough for instructions?

vincentqb commented on docs/source/optim.rst:
```diff
@@ -279,3 +279,30 @@ and start to collect SWA averages of the parameters at epoch 160:
+Functional Optimizer API
+---------------------------
+
+If you would like to update your parameters in a pure functional way, you can
+consider using the pure functional optimizer API. The functional APIs for
+optimizer allow you to only do the computation for the parameters you passed in,
+and inplace update the parameters accordingly.
```
simplifying, and incorporating some suggestions from alban

```suggestion
The computational part of the optimizers is contained in their respective functional APIs. They update only the parameters that were passed, and the updates are done inplace. They can also be used by users to perform a single step of an optimizer for a given state.
```

vincentqb commented on docs/source/optim.rst:
```diff
@@ -279,3 +279,30 @@ and start to collect SWA averages of the parameters at epoch 160:
+states before or after the call to this API (e.g. To use
+:func:`torch.optim.functional.adagrad`, `state_steps` for each parameter need
+to be updated before call to :func:`torch.optim.functional.adagrad`)
+
+As of now, we only exposed a limited set of functional APIs, we will
+expand the coverage and make all optimizers to have functional API in
+future releases.
```
Let's leave this in the warning below.

```suggestion
```

vincentqb commented on docs/source/optim.rst:
```diff
@@ -279,3 +279,30 @@ and start to collect SWA averages of the parameters at epoch 160:
+future releases.
+
+.. warning::
+    The functional optimizer API is still experimental and subject to change.
```
```suggestion
    The functional optimizer API is still experimental and subject to change. We are
    expanding the coverage of the functional APIs to all optimizers in future releases.
```

vincentqb commented on docs/source/optim.rst:
```diff
@@ -279,3 +279,30 @@ and start to collect SWA averages of the parameters at epoch 160:
+.. warning::
+    The functional optimizer API is still experimental and subject to change.
+
+
```
nit: one space?
```suggestion
```

vincentqb reviewed (changes requested):
I've made a few suggestions below, mostly about making it more concise. the main thing to add I would say is an example in the functional API itself on how to use it

Commit:
Update on "[doc] add functional optimizer to doc"


Add the basic documentation for functional optimizer to the doc

screenshot:

![Screen Shot 2020-09-29 at 11 36 40 AM](https://user-images.githubusercontent.com/9443650/94601283-1ea59100-0248-11eb-928b-61530bc04205.png)


Differential Revision: [D23995418](https://our.internmc.facebook.com/intern/diff/D23995418)

[ghstack-poisoned]

Commit:
Update on "[doc] add functional optimizer to doc"


Add the basic documentation for functional optimizer to the doc

screenshot:

![Screen Shot 2020-09-29 at 11 36 40 AM](https://user-images.githubusercontent.com/9443650/94601283-1ea59100-0248-11eb-928b-61530bc04205.png)


Differential Revision: [D23995418](https://our.internmc.facebook.com/intern/diff/D23995418)

[ghstack-poisoned]

vincentqb commented on docs/source/optim.rst:
```diff
@@ -279,3 +279,26 @@ and start to collect SWA averages of the parameters at epoch 160:
+call to :func:`torch.optim.functional.adagrad`).
+
+.. warning::
+    The functional optimizer API is still experimental and subject to change. We are
```
Let's double check the nomenclature "experimental", [internal doc](https://fb.quip.com/IOPWAxdbajaT#BDJACA1OHua).

vincentqb replied:
prototype is the correct name.

Commit:
Update on "[doc] add functional optimizer to doc"


Add the basic documentation for functional optimizer to the doc

screenshot:

![Screen Shot 2020-09-30 at 2 54 51 PM](https://user-images.githubusercontent.com/9443650/94743988-1ecb8c80-032d-11eb-9ac0-fb283134638e.png)


Differential Revision: [D23995418](https://our.internmc.facebook.com/intern/diff/D23995418)

[ghstack-poisoned]

vincentqb commented on docs/source/optim.rst:
```diff
@@ -279,3 +279,26 @@ and start to collect SWA averages of the parameters at epoch 160:
+call to :func:`torch.optim.functional.adagrad`).
+
+.. warning::
+    The functional optimizer API is a Beta level feature and subject to change. We are
```
@jspisak

vincentqb replied:
```suggestion
    The functional optimizer API is a prototype feature and subject to change. We are
```

Commit:
Update on "[doc] add functional optimizer to doc"


Add the basic documentation for functional optimizer to the doc

screenshot:

![Screen Shot 2020-09-30 at 2 54 51 PM](https://user-images.githubusercontent.com/9443650/94743988-1ecb8c80-032d-11eb-9ac0-fb283134638e.png)


Differential Revision: [D23995418](https://our.internmc.facebook.com/intern/diff/D23995418)

[ghstack-poisoned]

vincentqb commented on docs/source/optim.rst:
```diff
@@ -279,3 +279,26 @@ and start to collect SWA averages of the parameters at epoch 160:
+updates are done inplace. They can also be used by users to perform a single step
+of an optimizer for a given state.
+
+Note that you need to maintain the states of a functional optimizer by yourself.
+In particular, tensor states will be updated inplace inside the functional API, but
+you need to update some part of your non-tensor states before or after the call to
+this API (for instance, `state_steps` for each parameter need to be updated before
+call to :func:`torch.optim.functional.adagrad`).
```
This (`tensor states will be updated inplace inside the functional API, but
you need to update some part of your non-tensor states before or after the call to
this API`) and the above (`They update the parameters and the tensor states that were passed, and the
updates are done inplace.`) seem contradictory.

What is the exact contract of what is changed and what isn't?

vincentqb replied:
Would a code example help clarify?

vincentqb commented on docs/source/optim.rst:
```diff
@@ -279,3 +279,26 @@ and start to collect SWA averages of the parameters at epoch 160:
+
+The computational part of the optimizers is contained in their respective functional
+APIs. They update the parameters and the tensor states that were passed, and the
+updates are done inplace. They can also be used by users to perform a single step
```
nit:
```suggestion
updates are done in-place. They can also be used by users to perform a single step
```

vincentqb commented on docs/source/optim.rst:
```diff
@@ -279,3 +279,26 @@ and start to collect SWA averages of the parameters at epoch 160:
+of an optimizer for a given state.
+
+Note that you need to maintain the states of a functional optimizer by yourself.
+In particular, tensor states will be updated inplace inside the functional API, but
```
nit:
```suggestion
In particular, tensor states will be updated in-place inside the functional API, but
```

vincentqb commented:
relates to #45242
